libslide/src/evaluator_rules/unbuilt_rule.rs

```diff
@@ -0,0 +1,45 @@
+use crate::grammar::Expr;
+
+/// An unbuilt rule, generally used to express a rule in a human-readable form.
+#[derive(Copy, Clone)]
+pub enum UR {
+    /// An expression-mapping rule.
+    ///
+    /// A string rule is of the form
+    ///
+    /// ```text
+    /// "<expr> -> <expr>"
+    /// ```
+    ///
+    /// Where <expr> is any expression pattern. An expression pattern is similar to any other
+    /// expression, differing only in its pattern matching variables. The form of pattern matching
+    /// variables and the expressions they match are as follows:
+    ///   
+    /// | pattern | matches        |
+    /// |:------- |:-------------- |
+    /// | #<name> | A constant     |
+    /// | $<name> | A variable     |
+    /// | _<name> | Any expression |
+    ///
+    /// To apply a rule, the lhs of the rule is pattern matched on the target expression. If the
+    /// matching is sucessful, the rhs of the rule is expanded with the results of the matching.
+    ///
+    /// For example, the rule
+    ///   
+    /// ```text
+    /// "$a + 0 -> $a"
+    /// ```
+    ///
+    /// Applied on the expression `"x + 0"` would yield `"x"`.
+    ///
+    /// Note that mapping rules are built as, matched with, and applied on expression parse trees
+    /// rather than the string representations of expressions. This ensures rule application is
+    /// always exact and deterministic.
+    S(&'static str),
+
+    /// Multiple expression mappings representing one rule, perhaps in several different ways.
+    M(&'static [&'static str]),
+
+    /// A function rule.
+    F(fn(Expr) -> Option<Expr>),
+}
```

Comment:
These letters are confusing, I think we should expand on their naming. I didn't know what m at all intuitively until I read the comment.

Reply:
Can you also give an example of S v.s. M, I am not sure I understand.

Reply:
Yeah that's a good point. I was hoping the comments would be enough.

`M` is for when we need multiple patterns to express one rule. For example we can unwrap parentheses for either just constants or just variables, but we have no pattern that describes constant|variable, so we have to use two patterns -- `"($a) -> $a"` and `"(#a) -> #a"`.

Reply:
gotcha.